model/train.py

```diff
@@ -397,60 +397,123 @@ def evaluate(*, p_eval_step, target, eval_ds: tf.data.Dataset,
 def decode_and_calculate_acc(*, p_pred_step, p_init_cache, target, config,
                                  predict_ds: tf.data.Dataset, decode_tokens,
                                  encode_tokens, max_predict_length: int, 
-                                 max_predict_loops = 1, use_annotations = False,
-                                 extra_loops = 0):
+                                 max_predict_loops=1, use_annotations=False,
+                                 extra_loops=0, copy_input=False,  
+                                 copy_input_in_full=False, copy_output=False,
+                                 end_token=None, in_out_token=None,
+                                 sep_token=None, end_iter_token=None):
   """Processes the `predict_ds` and calculates the sentence accuracy from 
   decoded predictions."""
   n_devices = jax.local_device_count()
   logging.info("Translating evaluation dataset.")
   sources, references, predictions = [], [], []
   wrong_preds, wrong_refs, wrong_sources = [], [], []
   predicted_list_all_batches = []
+  
   for pred_batch in predict_ds:
     pred_batch = jax.tree_map(lambda x: x._numpy(), pred_batch)  # pylint: disable=protected-access
     # Handle final odd-sized batch by padding instead of dropping it.
     cur_pred_batch_size = pred_batch["inputs"].shape[0]
+    per_dev_batch_size = int(cur_pred_batch_size / n_devices)
     if cur_pred_batch_size % n_devices:
-      padded_size = int(np.ceil(cur_pred_batch_size / n_devices) * n_devices)
+      per_dev_batch_size = int(np.ceil(cur_pred_batch_size / n_devices))
+      padded_size = int(per_dev_batch_size * n_devices)
       pred_batch = jax.tree_map(
           lambda x: pad_examples(x, padded_size),  # pylint: disable=cell-var-from-loop
           pred_batch)
     pred_batch = common_utils.shard(pred_batch)
     cache = p_init_cache(pred_batch["inputs"])
-    count = 0
+    
     predicted = p_pred_step(pred_batch['inputs'], target, cache, decode.EOS_ID,
                             max_predict_length)
-    predicted_list = [tohost(predicted)]
+    concatenated_predicted = predicted
+    predicted_list = [tohost(concatenated_predicted)]
+    this_input = pred_batch["inputs"]
+    
     if use_annotations:
         max_predict_loops = jnp.max(pred_batch['op']) + extra_loops
+        
+    count = 0
     while count < max_predict_loops - 1: 
+        if copy_input:
```

Comment:
This function is getting fairly large. Any chance of cutting any piece out with an auxiliary function for clarity?